### PR TITLE
sql: `pg_proc` and `SHOW CREATE FUNCTION`

### PR DESCRIPTION
**Pull request by chengxiong-ruan**
There 3 commits:
(1) populate `pg_proc` with UDF
(2) create `crdb_internal.create_function_statements` virtual table
(3) implement `SHOW CREATE FUNCTION`
(4) add `::regproc` support for UDFs

### CODE REVIEW

File: pkg/sql/catalog/typedesc/type_desc.go
Comment (chengxiong-ruan) [inline pkg/sql/catalog/typedesc/type_desc.go:799]: I think this was a bug that `EnsureTypeIsHydrated` didn't hydrate table implicit types which is actually a tuple. Let me know if this make sense..

Comment (rafiss) [reply pkg/sql/catalog/typedesc/type_desc.go:799]: if we make this change, let's also add a test that shows why it's needed.

**Review by rafiss — CHANGES_REQUESTED** [review]
i just had some small nits. nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-:-N8izz4s3mlX4ztHh5Ep:bejp4y9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @chengxiong-ruan)

___
*[`pkg/sql/crdb_internal.go` line 2581 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8ivnZF6G4I6zeSTKTd:-N8ivnZF6G4I6zeSTKTe:bhma2yz) ([raw file](https://github.com/cockroachdb/cockroach/blob/985487f3924b5fdae40f8c592c827b0fb2978d23/pkg/sql/crdb_internal.go#L2581)):*
> ```Go
> 
> var crdbInternalCreateFunctionStmtsTable = virtualSchemaTable{
> 	comment: "CREATE statements for all ",
> ```

nit: incomplete sentence
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8iwFfDDEhw_KbIZXi4:-N8iwFfDDEhw_KbIZXi5:b678ie6) ([raw file](https://github.com/cockroachdb/cockroach/blob/985487f3924b5fdae40f8c592c827b0fb2978d23/pkg/sql/catalog/funcdesc/func_desc.go#L538)):*
> ```Go
> 
> // ToTreeNode implements the FunctionDescriptor interface.
> func (desc *immutable) ToTreeNode() (ret *tree.CreateFunction, err error) {
> ```

nit: i think a name like `func (desc *immutable) CreateExpr` or `func (desc *immutable) ToCreateExpr` might be a bit more natural
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 570 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8iwx9E8w8PpqejbEuz:-N8iwx9E8w8PpqejbEv-:b-nfubsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/985487f3924b5fdae40f8c592c827b0fb2978d23/pkg/sql/catalog/funcdesc/func_desc.go#L570)):*
> ```Go
> }
> 
> func (desc *immutable) getTreeNodeLang() tree.FunctionLanguage {
> ```

i think the `TreeNode` part of the name is slightly redundant, since we can already see that in the return type
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 578 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8ix60B7DnePv944DnJ:-N8ix60B7DnePv944DnK:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/985487f3924b5fdae40f8c592c827b0fb2978d23/pkg/sql/catalog/funcdesc/func_desc.go#L578)):*
> ```Go
> }
> 
> func (desc *immutable) getTreeNodeVolatility() tree.FunctionVolatility {
> ```

ditto
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 590 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8ix7YR-U_oAQW_erKm:-N8ix7YR-U_oAQW_erKn:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/985487f3924b5fdae40f8c592c827b0fb2978d23/pkg/sql/catalog/funcdesc/func_desc.go#L590)):*
> ```Go
> }
> 
> func (desc *immutable) getTreeNodeNullInputBehavior() tree.FunctionNullInputBehavior {
> ```

ditto

File: pkg/sql/catalog/typedesc/type_desc.go
Comment (chengxiong-ruan) [reply pkg/sql/catalog/typedesc/type_desc.go:799]: ah yeah, good call, I'll add one for it.

**Comment by chengxiong-ruan** [comment]
@rafiss sorry I didn't see your reviews and I added an extra small commit to add `::regproc` support for UDFs. I'll address your earlier comments. Thanks.

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85656)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @chengxiong-ruan)

___
*[`pkg/sql/sem/eval/parse_doid.go` line 73 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8jA8OP82R8q-fG9ydO:-N8jA8OP82R8q-fG9ydP:b-4s6gz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d076135e4f93da8b8c8d7496ec8ec78e34a392dc/pkg/sql/sem/eval/parse_doid.go#L73)):*
> ```Go
> 			name.Parts[i] = substrs[len(substrs)-1-i]
> 		}
> 		funcDef, err := ctx.Planner.ResolveFunction(ctx.Ctx(), &name, &ctx.SessionData().SearchPath)
> ```

you will also need to update `performIntToOidCast` inside eval/cast.go (and add tests please)

**Review by chengxiong-ruan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85656)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @rafiss)

___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8iwFfDDEhw_KbIZXi4:-N8jTLV40N5lY5MXgvk6:bxy3e87) ([raw file](https://github.com/cockroachdb/cockroach/blob/985487f3924b5fdae40f8c592c827b0fb2978d23/pkg/sql/catalog/funcdesc/func_desc.go#L538)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: i think a name like `func (desc *immutable) CreateExpr` or `func (desc *immutable) ToCreateExpr` might be a bit more natural
</blockquote></details>

yeah, good point, fixed
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 570 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8iwx9E8w8PpqejbEuz:-N8jUSgV9xPvjZGvl4eM:b3jhwrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/985487f3924b5fdae40f8c592c827b0fb2978d23/pkg/sql/catalog/funcdesc/func_desc.go#L570)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think the `TreeNode` part of the name is slightly redundant, since we can already see that in the return type
</blockquote></details>

yeah, but I want to distinguish from the function return the `catpb` one. Inspired by your another comment on naming, I replaced `TreeNode` with `CreateExpr`. Hope that makes sense.
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 578 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8ix60B7DnePv944DnJ:-N8jVGMk7GQt8X9wcHMD:biho5l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/985487f3924b5fdae40f8c592c827b0fb2978d23/pkg/sql/catalog/funcdesc/func_desc.go#L578)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

fixed with same renaming
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 590 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8ix7YR-U_oAQW_erKm:-N8jVJSS6t0aP7SZ-0pL:biho5l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/985487f3924b5fdae40f8c592c827b0fb2978d23/pkg/sql/catalog/funcdesc/func_desc.go#L590)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

fixed with same renaming
___
*[`pkg/sql/sem/eval/parse_doid.go` line 73 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8jA8OP82R8q-fG9ydO:-N8jVLC9-5YAFd1w7DyT:b-p0u6cz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d076135e4f93da8b8c8d7496ec8ec78e34a392dc/pkg/sql/sem/eval/parse_doid.go#L73)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

you will also need to update `performIntToOidCast` inside eval/cast.go (and add tests please)
</blockquote></details>

fixed that, and also added a test to make sure function using table implicit types are hydrated properly.

**Review by ajwerner — APPROVED** [review]
LGTM

**Comment by chengxiong-ruan** [comment]
TFTR!
bors r+

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85656)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @rafiss)

___
*[`pkg/sql/sem/eval/cast.go` line 943 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8xviQS81RuIn5pkyVA:-N8xviQS81RuIn5pkyVB:b4xax) ([raw file](https://github.com/cockroachdb/cockroach/blob/381204878a149f62bb24b9c99fc96ad503c5393d/pkg/sql/sem/eval/cast.go#L943)):*
> ```Go
> 
> 	case oid.T_regproc, oid.T_regprocedure:
> 		name, _, err := res.ResolveFunctionByOID(ctx, oid.Oid(v))
> ```

i think this is going to have the same issue that was just fixed in https://github.com/cockroachdb/cockroach/pull/85086 -- it's not always safe to ignore this error. (specifically, in the case where the error causes the txn to abort.) so i think you may want something similar where `ResolveFunctionByOID` also returns a "safe to ignore error" flag. OR refactor things so that this code never needs to ignore the error.

**Comment by chengxiong-ruan** [comment]
*[`pkg/sql/sem/eval/cast.go` line 943 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8xviQS81RuIn5pkyVA:-N8xxsS4EPlcsCHbaCyM:bdi4j6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/381204878a149f62bb24b9c99fc96ad503c5393d/pkg/sql/sem/eval/cast.go#L943)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think this is going to have the same issue that was just fixed in https://github.com/cockroachdb/cockroach/pull/85086 -- it's not always safe to ignore this error. (specifically, in the case where the error causes the txn to abort.) so i think you may want something similar where `ResolveFunctionByOID` also returns a "safe to ignore error" flag. OR refactor things so that this code never needs to ignore the error.
</blockquote></details>

yeah, good point, I added a `ErrFunctionUndefined` in [another pr](https://github.com/cockroachdb/cockroach/pull/85718) which bases on this pr. I'll fix this separately outside of this PR, does that sound good?

**Review by rafiss — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85656)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @rafiss)

___
*[`pkg/sql/sem/eval/cast.go` line 943 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/85656#-N8xviQS81RuIn5pkyVA:-N8xz_-9536Eh9JMfySN:bmphp0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/381204878a149f62bb24b9c99fc96ad503c5393d/pkg/sql/sem/eval/cast.go#L943)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

yeah, good point, I added a `ErrFunctionUndefined` in [another pr](https://github.com/cockroachdb/cockroach/pull/85718) which bases on this pr. I'll fix this separately outside of this PR, does that sound good?
</blockquote></details>

sure that sounds good

**Comment by chengxiong-ruan** [comment]
TFTR!
bors r+

**Comment by chengxiong-ruan** [comment]
need to rebased and update udf logic test with increased ids since there were new objects added.
bors r-

**Comment by chengxiong-ruan** [comment]
only one flaky test failed, thanks again for the reviews!
bors r+

**Comment by chengxiong-ruan** [comment]
canceling to wait until Marcus' #85674 is merged, there is conflict.
bors r-

**Comment by chengxiong-ruan** [comment]
TFTR again!
bors r+

**Comment by abarganier** [comment]
Posting here for visibility - I think we have a nil pointer issue with the crdb_internal table when querying across databases. 

Issue here: https://github.com/cockroachdb/cockroach/issues/85897

**Pull request merged**
